chain/chunks/src/lib.rs

```diff
@@ -1357,20 +1465,105 @@ impl ShardsManager {
+        if let Some(forwarded_parts) = self.chunk_forwards_cache.cache_remove(&chunk_hash) {
+            // We have the header now, and there were some parts we were forwarded earlier.
+            // Let's process those parts now.
+            let forwarded_chunk = PartialEncodedChunkV2 {
+                header: header.clone(),
+                parts: forwarded_parts.into_iter().map(|(_, part)| part).collect(),
+                receipts: Vec::new(),
+            };
+            // Make sure we mark this chunk as being requested so that it is handled
+            // properly in the next call. Note no requests are actually sent at this point.
+            self.request_chunk_single_mark_only(header);
+            // Call process_partial_encoded_chunk recursively, "simulating" as that forwarded
+            // part is just received from the network
+            return self.process_partial_encoded_chunk(
+                // We can assert the signature on the header is valid because
+                // it would have been checked in an earlier call to this function.
+                MaybeValidated::from_validated(&forwarded_chunk),
+                None,
+                chain_store,
+                rs,
+            );
+        }
+
+        // 4. Forward my parts to others tracking this chunk's shard
+        let epoch_id = match self
+            .runtime_adapter
+            .get_epoch_id_from_prev_block(&partial_encoded_chunk.header.prev_block_hash())
+        {
+            Ok(epoch_id) => epoch_id,
+            Err(_) => {
+                return Ok(ProcessPartialEncodedChunkResult::NeedBlock);
+            }
+        };
+        let protocol_version = self.runtime_adapter.get_epoch_protocol_version(&epoch_id)?;
         if checked_feature!("stable", ForwardChunkParts, protocol_version) {
             self.send_partial_encoded_chunk_to_chunk_trackers(partial_encoded_chunk)?;
         };
 
-        let entry = self.encoded_chunks.get(&chunk_hash).unwrap();
+        self.check_chunk_have_all_parts_and_receipts(&partial_encoded_chunk.header, chain_store, rs)
+    }
+
+    /// Checks if the chunk has all parts and receipts, if so and if the node cares about the shard,
+    /// decode and persist the full chunk
+    /// `header`: header of the chunk. It must be known by `ShardsManager`, either
+    ///           by previous call to `process_partial_encoded_chunk` or `request_partial_encoded_chunk`
+    pub fn check_chunk_have_all_parts_and_receipts(
+        &mut self,
+        header: &ShardChunkHeader,
+        chain_store: &mut ChainStore,
+        rs: &mut ReedSolomonWrapper,
+    ) -> Result<ProcessPartialEncodedChunkResult, Error> {
+        // The logic from now on requires previous block is processed because
+        // calculating owner parts requires that
+        let prev_block_hash = header.prev_block_hash();
+        let epoch_id = match self.runtime_adapter.get_epoch_id_from_prev_block(&prev_block_hash) {
```

Comment:
seems that we're already doing it before calling (in line 1495) -- maybe we should change this method to receive epoch_id in arguments?

Reply:
It has to be checked in this function because it is not just for getting epoch_id, it is also to check the previous block has been processed by the chain. The functions `has_all_parts` requires that